File: uploader/src/main/java/net/hearthstats/ui/ClickableDeckBox.java
package net.hearthstats.ui;

import java.awt.Dimension;
import java.awt.event.MouseAdapter;
import java.awt.event.MouseEvent;
import java.io.File;
import java.io.FileOutputStream;
import java.net.URL;
import java.nio.channels.Channels;
import java.nio.channels.ReadableByteChannel;
import java.util.concurrent.Executors;
import java.util.concurrent.ScheduledExecutorService;
import java.util.concurrent.TimeUnit;

import javax.swing.Box;
import javax.swing.ImageIcon;
import javax.swing.JLabel;
import javax.swing.SwingUtilities;

import net.hearthstats.Card;
import net.hearthstats.Deck;
import net.hearthstats.log.Log;

public class ClickableDeckBox {

	private ClickableDeckBox() {
	}

	public static Box makeBox(Deck deck) {
		downloadImages(deck);
		Box container = Box.createHorizontalBox();
		Box box = Box.createVerticalBox();
		JLabel imageLabel = new JLabel();
		imageLabel.setPreferredSize(new Dimension(289, 398));
		for (Card card : deck.cards()) {
			ClickableLabel cardLabel = new ClickableLabel(card);
			box.add(cardLabel);
			cardLabel.addMouseListener(new MouseHandler(card, imageLabel));
		}
		container.add(box);
		container.add(imageLabel);
		return container;
	}

	private static void downloadImages(Deck deck) {
		ScheduledExecutorService scheduledExecutorService = Executors
				.newScheduledThreadPool(30);

		for (final Card card : deck.cards()) {
			scheduledExecutorService.submit(new Runnable() {
				@Override
				public void run() {
					try {
						ReadableByteChannel rbc = Channels.newChannel(new URL(
								card.url()).openStream());
						File file = card.localFile();
						if (file.length() < 30000) {
							// probably not downloaded correctly yet
							FileOutputStream fos = new FileOutputStream(file);
							fos.getChannel().transferFrom(rbc, 0,
									Long.MAX_VALUE);
							fos.close();
							rbc.close();
							Log.debug(card.name() + " saved to cache folder");
						} else {
							Log.debug(card.name()
									+ " already in cache, skipping");
						}
					} catch (Exception e) {
						Log.error(
								"Could not download image for " + card.name(),
								e);
					}
				}
			});
		}
		try {
			scheduledExecutorService.shutdown();
			scheduledExecutorService.awaitTermination(30, TimeUnit.SECONDS);
			Log.info("all images downloaded successfully");
		} catch (InterruptedException e) {
			throw new RuntimeException(e);
		}
	}

	private static class MouseHandler extends MouseAdapter {
		private Card card;
		private JLabel imageLabel;

		public MouseHandler(Card card, JLabel imageLabel) {
			this.card = card;
			this.imageLabel = imageLabel;
		}

		@Override
		public void mouseEntered(MouseEvent e) {
			SwingUtilities.invokeLater(new Runnable() {
				@Override
				public void run() {
					imageLabel.setIcon(new ImageIcon(card.localURL()));
				}
			});
		}
	}

}

File: uploader/src/main/scala/net/hearthstats/ui/ClickableDeckBox.scala
package net.hearthstats.ui

import java.awt.Dimension
import java.awt.event.MouseAdapter
import java.awt.event.MouseEvent
import scala.swing.Swing.onEDT
import javax.swing.Box.createVerticalBox
import javax.swing.BoxLayout
import javax.swing.ImageIcon
import javax.swing.JLabel
import javax.swing.JPanel
import net.hearthstats.Card
import net.hearthstats.CardUtils
import net.hearthstats.Deck
import net.hearthstats.log.Log
import rx.lang.scala.Observable
import net.hearthstats.logmonitor.CardEvent
import net.hearthstats.logmonitor.CardEventType._
import javax.swing.JFrame
import scala.swing.Frame
import scala.swing.MainFrame
import scala.swing.BorderPanel
import java.awt.event.WindowAdapter
import java.awt.event.WindowEvent
import javax.swing.WindowConstants
import rx.lang.scala.observables.ConnectableObservable
import net.hearthstats.OldConfig
import java.awt.BorderLayout
import javax.swing.JCheckBox
import scala.swing.Swing._
import scala.concurrent.Future
import net.hearthstats.config.EnvironmentComponent
import ConfigComponent

trait DeckOverlaySwingComponent extends DeckOverlayPresenterComponent { self: ConfigComponent =>
  val deckOverlay = new DeckOverlaySwing

  class DeckOverlaySwing extends JFrame with DeckOverlayPresenter {
    import config._

    var cardLabels: Map[String, ClickableLabel] = Map.empty

    def showDeck(deck: Deck) {
      val imagesReady = CardUtils.downloadImages(deck.cards)

      val content = getContentPane
      content.setLayout(new BorderLayout)
      val box = createVerticalBox
      val imageLabel = new JLabel
      imageLabel.setPreferredSize(new Dimension(289, 398))
      cardLabels =
        (for {
          card <- deck.cards
          cardLabel = new ClickableLabel(card, imagesReady)
        } yield {
          box.add(cardLabel)
          cardLabel.addMouseListener(new MouseHandler(card, imageLabel))
          card.name -> cardLabel
        }).toMap

      val cardCheckBox = new JCheckBox("Display full card image")
      cardCheckBox.addChangeListener(ChangeListener(_ => imageLabel.setVisible(cardCheckBox.isSelected)))
      content.add(cardCheckBox, BorderLayout.NORTH)
      content.add(box, BorderLayout.CENTER)
      content.add(imageLabel, BorderLayout.EAST)
      imageLabel.setVisible(false)

      setAlwaysOnTop(true)
      setFocusableWindowState(true)
      setDefaultCloseOperation(WindowConstants.DISPOSE_ON_CLOSE)

      setLocation(deckX, deckY)
      setSize(deckWidth, deckHeight)
      setVisible(true)

    }

    override def dispose(): Unit = {
      //    connection.unsubscribe()
      try {
        // Save the location of the window, if it is visible (it won't be visible if the selected deck is invalid)
        if (isVisible) {
          val p = getLocationOnScreen
          deckX.set(p.x)
          deckY.set(p.y)
          val rect = getSize()
          deckWidth.set(rect.getWidth.toInt)
          deckHeight.set(rect.getHeight.toInt)
        }
      } catch {
        case e: Exception =>
          Log.warn("Error occurred trying to save your settings, your deck overlay position may not be saved", e)
      }
      super.dispose()
    }

    //  val newEvents = cardEvents.publish
    //  val connection = newEvents.connect
    //  newEvents.subscribe {
    //    _ match {
    //      case CardEvent(card, DRAWN) => findLabel(card) map (_.decreaseRemaining())
    //      case CardEvent(card, REPLACED) => findLabel(card) map (_.increaseRemaining())
    //    }
    //  }

    private def findLabel(c: Card): Option[ClickableLabel] =
      if (c.collectible) {
        val l = cardLabels.get(c.name)
        if (l.isDefined) l
        else {
          Log.warn(s"card ${c.name} not found in deck")
          None
        }
      } else None

    def removeCard(card: Card)

    def addCard(card: Card)

    case class MouseHandler(card: Card, imageLabel: JLabel) extends MouseAdapter {
      override def mouseEntered(e: MouseEvent) {
        onEDT(imageLabel.setIcon(new ImageIcon(card.localURL)))
      }
    }
  }
}
